### ISSUE TITLE
C++ 20

### ISSUE THREAD

**Issue by Woazboat** [zerebubuth/openstreetmap-cgimap#323]
What's the status for moving from C++17 to C++20?

Is it planned to do so eventually? Are there any blockers?

**Comment by mmd-osm** [comment]
I haven't really thought about this in detail. I'd say it depends a bit on the libaries we're using. libpqxx as an example announced that the next 8.x version requires C++20. However, it might take some time until this shows up as a package in the Debian releases we're using. 

Were you looking for some specific features which would make our lifes easier with C++20 today? One candidate could be https://en.cppreference.com/w/cpp/utility/format to remove the fmt dependency. Anything else maybe?

**Comment by Woazboat** [comment]
Also things like this https://github.com/zerebubuth/openstreetmap-cgimap/pull/280#issuecomment-1325238300

**Comment by tomhughes** [comment]
You need gcc 13 to get `std::format` though so Ubuntu 22.04 won't be able to use it even if you use C++20 mode.

I'd forget about coroutines as well - they're more theoretical than practical in C++20.

**Comment by Woazboat** [comment]
Vanilla coroutines in c++20 are unfortunately not very usable, yes. 
Anyway, there are no downsides to migrating as far as I can see and it'll be needed at some point in the future anyway.

**Comment by Woazboat** [comment]
Okay, found blocker # 1

The libpqxx packages version < 7.9.0 as shipped in Debian (and very likely Ubuntu as well) are built with C++17 and cannot be linked with programs built with C++20

https://github.com/jtv/libpqxx/issues/732

**Comment by tomhughes** [comment]
That's odd - there's no ABI break in libstdc++ for C++20 at all. The last ABI break was for `std::string` and `std::list` in C++11.

Looking at that ticket it looks like there is indeed something using the pre-C++11 `std::string` ABI for some reason - technically switching that ABI is not tied to the C++ version as such but I'm not sure when Debian/Ubuntu switched the default though Fedora did it long ago.

**Comment by mmd-osm** [comment]
Based on https://stackoverflow.com/questions/75066606/indentation-aware-raw-string-literals we could remove leading spaces in the raw string literals used in SQL statements: https://godbolt.org/z/5ovaq8E91

**Comment by pnorman** [comment]
Is the latest post in the right place? It seems disconnected for from C++20

**Comment by mmd-osm** [comment]
Yes, it's related to C++20. Once you switch to something pre C++20 in godbolt, it will no longer compile.

**Comment by Woazboat** [comment]
> Okay, found blocker # 1
> 
> The libpqxx packages version < 7.9.0 as shipped in Debian (and very likely Ubuntu as well) are built with C++17 and cannot be linked with programs built with C++20
> 
> [jtv/libpqxx#732](https://github.com/jtv/libpqxx/issues/732)


I looked into this a bit more to understand what's actually going on there. It's not actually a libstdc++ ABI break issue, the code is already using the post C++11 ABI version.

Function definitions in libpqxx < v7.9.0 include headers are defined incorrectly in a way that causes them to vary depending on available language features / C++ standard version. This causes a mismatch between the compiled code in the installed shared library binaries and the development headers.

For C++ standard versions after C++ 20, pqxx exceptions in except.hxx / except.cxx have an additional std::source_location constructor parameter and class member that are not present for older C++ versions. The libpqxx shared library binaries distributed in Debian and Ubuntu are compiled with an older C++ standard version, so the compiled code does not have the additional std::source_location member / constructor parameter. 
However, the development headers distributed alongside the shared library are not fixed to the C++ version the library was compiled with and have different function / class declarations depending on the C++ standard version that is used to compile the code that includes these headers. This causes a mismatch between the declarations in the header files and the actual pre-compiled code in the shared library binary when compiling with C++ 20. The code using the library sees the function declarations with the additional std::source_location parameters, but the library actually only contains the versions without the additional parameter.

**Comment by mmd-osm** [comment]
Someone posted this workaround in the linked issue:

```
// hack to fix building under Linux
#ifdef __cpp_lib_source_location
#undef __cpp_lib_source_location
#endif
#define __cpp_lib_source_location 0

#include <pqxx/pqxx>
```

I haven’t tried this yet…

**Comment by Woazboat** [comment]
https://packages.debian.org/search?keywords=libpqxx
Debian stable still ships libpqxx version `6.4` and testing/unstable has since updated to version `7.10` which is no longer affected.

https://launchpad.net/ubuntu/+source/libpqxx
Ubuntu 24.04 LTS unfortunately ships the affected version `7.8`, but the current non-LTS stable release 24.0 should already have the fixed version `7.9`.

**Comment by mmd-osm** [comment]
Right, Ubuntu 24.04 is the only distro where this is causing issues at the moment. I'm using a static build of pqxx 7.10.0 to work around the issue as temporary workaround only: https://github.com/mmd-osm/openstreetmap-cgimap/blob/cpp20_1/docker/ubuntu/Dockerfile2404#L18-L27

cpp20_1 is more of an experimental branch to try out some new features and see what breaks. Likely some of that material will show up in a different form in future PRs.

**Comment by Woazboat** [comment]
@tomhughes is compatibility with Ubuntu 24.04 something that needs to be considered for deployment or are the servers all Debian now anyway?

**Comment by tomhughes** [comment]
All the production servers running cgimap are Debian 12 now.
